### DISCUSSION TITLE
git lfs is not working

### DISCUSSION THREAD

**Discussion by ryan-d-p-ubana** [git-lfs/git-lfs#5817] [Q&A]
I have tried to install git lfs (windows) in the cloned repository but it is not working. It always displays the error below: But when I remove the .git from the cloned repository and tried to git lfs install it is working. How to resolve this issue?
```
Command: git lfs install
Command: git lfs track <large file>
Command: git add . gitattributes
Command: git commit -m "files"
Command: git push origin main

Total 21 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), done.
remote: error: Trace: 41bdc94df8f8c1b9054b15c2c7e2623c46b048c7d113353173addbb38525095b
remote: error: See https://gh.io/lfs for more information.
remote: error: File oc is 124.14 MB; this exceeds GitHub's file size limit of 100.00 MB
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
To https://github.com/PMTJ-roks/ocp_object_template.git
 ! [remote rejected] main -> main (pre-receive hook declined)


```

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.

It sounds like you may have some pre-existing files in your Git commit history which are already over GitHub's size limit of 100 MB.  If that's the case, you'll likely need to rewrite your Git repository's history so that every instance of the large files in every existing commit is converted into an object tracked by Git LFS.  Unfortunately, just running `git lfs track` isn't sufficient to do this, as that will only help with _new_ files you create in subsequent commits which match the patterns you provide to `git lfs track`.

(One other note: Even if you're just adding new files, you'll want to run `git add .gitattributes`, not `git add . gitattributes`.  The file is named `.gitattributes`; the latter command would instead add all files in the current directory, i.e., the `.` directory.)

To be clear about what I mean as a pre-existing file, I'm referring to anything you've already committed into your local Git history, even if you haven't pushed that Git history to GitHub or another remote yet.

If you haven't pushed anything yet, and have the option to unwind your local Git commits with the large file (that is, delete them and recreate them from scratch), you could do that with commands like `git reset --hard HEAD^` (which deletes the most recent commit), then run `git lfs track` _before_ adding the large files into your Git history.  That's obviously a destructive approach, though, so if you have a complex commit you won't be able to easily recreate, follow the migration steps below instead.

Assuming you have pre-existing large files in your Git commit history you need to migrate to Git LFS, the `git lfs migrate import` command will help.  The [manual page](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/docs/man/git-lfs-migrate.adoc) describes the options for that command.  You probably want to run something like the following (after you've run `git lfs install`, of course, but I believe you've done that already):

```
$ git lfs migrate import --everything --include="<large file pattern>" 
```

The `--everything` means to process all branches and tags, not just the current branch.

Once this has run, your Git history will have been changed so that any files matching the pattern you provide in `--include` are converted to Git LFS pointers and objects in all commits in your history, and appropriate `.gitattributes` patterns are introduced at the same time in your history.

Now it should be possible to push those branches and tags to GitHub.  However, note that if there were smaller files earlier in your Git commit history which also match the pattern you supplied, they'll have been converted to Git LFS pointers and objects too.  For instance, if you ran `git lfs migrate import --everything --include="*.jpg"` then every `*.jpg` file in your history will now be a Git LFS object.  If you've previously pushed commits with those files to your remote, then you'll need to "force-push" to overwrite those commits on the remote.  This may not be an issue for you, if you're just dealing with one new file which has never been pushed to GItHub and you're using a file pattern which matches it exclusively.  But if you're using a wildcard pattern like `*.jpg`, and that matches files in your Git history you've previously pushed to your remote, you'll need to push with `--force`.  Note, again, that this is a destructive process -- your old history will be overwritten on the server -- so please be certain this is OK and that the rewritten history looks correct and acceptable to you.

**Comment by ryan-d-p-ubana** [comment]
Hi @chrisd8088 ,
Thank you for your great reply. Currently I am using BFG repo cleaner is it the same as git filter? I am using it to remove large commits in the history. But regarding converting to lfs. Is it possible to use BFG repo cleaner as well when dealing LFS? Also, I am using git clone --bare command to clone the repository. But I think git lfs wont work with the cloned repo using git clone --bare? 

My concern is how can I import LFS to GitHub Enterprise without losing the commits, branches, and tag when being pushed to github?

**Reply by chrisd8088** [reply]
I'm not too familiar with BFG, so I won't try to provide advice on that score.

I'm also a little unclear on what your process is here.  Do you mean that you've made a bar clone, and then want to run `git lfs migrate import` or some other kind of filtering command in that clone?  That should work OK with `git lfs migrate import`, I believe.

**Comment by ryan-d-p-ubana** [comment]
Is there any specific command for it? Because those are important to retain in the project. Another question is it safe to use BFG repo cleaner to remove lfs history and converting to git lfs? what is the best approach ?

**Comment by ryan-d-p-ubana** [comment]
Also this command is working when working with LFS Command:  git lfs migrate import --everything --include="<lfs-file>".
But I noticed when I used the command: git push github --force is that other branches and tags were not included. How it is possible to push other branches and tags using single command?

**Reply by chrisd8088** [reply]
You probably want to use `git push --force --all ...` to push your other branches and tags besides your currently checked-out branch.

**Comment by ryan-d-p-ubana** [comment]
Hi @chrisd8088 ,

The command  git lfs migrate import --everything --include="<lfs-files>" is working. I also used git clone --bare and then git push --mirror. Regarding the command is it safe to use when I have LFS files and large commit history? Also, it wont remove the physical file? I am able to use it independently without the use of git lfs track "lfs-file"

**Reply by chrisd8088** [reply]
Yes, `git lfs migrate import` is safe to use with a large commit history; it may just take more time to run.  No data should be lost.  However, see my comments below about the notion of "safety".

No, it won't remove "physical" files.  It's worth noting that Git doesn't represent individual blobs in the commit history as individual files, though -- they will be packed into its internal file format along with other blobs, trees, and commits.  What Git LFS migration will is replace the full blob in the Git history with a pointer blob, and then place the full blob's contents into Git LFS's own internal storage cache under `.git/lfs/objects`.

Yes, `git lfs migrate import` and `git lfs track` can—and indeed, often should—be used together.  Migration is necessary if you have blobs in your commit history you need to convert to Git LFS.  That will also insert the appropriate patterns into `.gitattributes` in your commit history too.  The `git lfs track` command is really just a helper which also add Git LFS patterns to `.gitattributes`; you can do this manually, though, if you know what to add.  Those patterns will then apply to any new blobs created in the future.

Note that you may not need to run `git lfs track` after running `git lfs migrate import`; it depends on what patterns you're adding.  For example, if you run `git lfs migrate import --include="*.jpg"`, then you probably don't also need to run `git lfs track "*.jpg"` because the `*.jpg` pattern should already be in your `.gitattributes` after the migration step.  If you run `git lfs track "*.jpg"` you'd likely just see `"*.jpg" already supported`.

To come back to the topic of "safety", I do want to emphasize that there is a subjective aspect to this concept as well as an objective one.  No Git or Git LFS command should unexpectedly delete or destroy content.  If they do, that's a bug.  So in that objective sense, `git lfs migrate import` should never lose content, and is therefore safe to use.

But if you don't understand what the command does, and how it operates and changes your Git history, and how Git blobs and Git LFS objects are stored, it may seem like things have gone awry and data has disappeared.  So it's very important to have an understanding of how Git and Git LFS store data before you proceed to migrate a large, complex commit history to Git LFS.

I would recommend [reading](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects) up on how Git stores its data, and some of Git LFS's documentation like the pointer [specification](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/docs/spec.md) and our wiki's [tutorial](https://github.com/git-lfs/git-lfs/wiki/Tutorial) and our FAQ [manual page](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/docs/man/git-lfs-faq.adoc).

**Comment by ryan-d-p-ubana** [comment]
Hi @chrisd8088 ,

Im trying to push to github a size of 2.26GB from Gitlab export. Unfortunately, I encountered an error below:
error: RPC failed; HTTP 500 curl 22 The requested URL returned error: 500
fatal: the remote end hung up unexpectedly


Username for 'https://github.com': Ryan.Danielle.Ubana@ibm.com
Password for 'https://Ryan.Danielle.Ubana@ibm.com@github.com':
Enumerating objects: 163334, done.
Counting objects: 100% (163334/163334), done.
Delta compression using up to 8 threads
Compressing objects: 100% (33999/33999), done.
error: RPC failed; HTTP 500 curl 22 The requested URL returned error: 500
fatal: the remote end hung up unexpectedly
Writing objects: 100% (163334/163334), 2.26 GiB | 6.68 MiB/s, done.
Total 163334 (delta 99019), reused 163334 (delta 99019)
fatal: the remote end hung up unexpectedly
Everything up-to-date

Do you have solution to this?

**Reply by chrisd8088** [reply]
First, since you're using GitHub, you can check with GitHub Support for additional assistance; this project is just for the open-source Git LFS client, so we can't always help diagnose problems with specific Git LFS hosting services as well as their support teams can.

Second, try running your commands after settings these environment variables to enable tracing: `GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1`.  For instance, you might run:

```
$ GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 git push origin main >push.log 2>&1
```

Then take a look at the `push.log` file.

Third, I notice there's this Stack Overflow [question](https://stackoverflow.com/questions/15240815/git-fatal-the-remote-end-hung-up-unexpectedly) which suggests you may want to set the `http.postBuffer` Git configuration value to a larger value (`git config http.postBuffer ...`).  Again, though, I would check with GitHub's Support team, and if you send them the `push.log` trace you create, I suspect they will be appreciative of that.

**Reply by KoxyG** [reply]
Hi, after successfully pushing my video file to github using lfs, the video file is not working anymore. How can this be fix so my video can display.

**Reply by chrisd8088** [reply]
@KoxyG -- please open a new [discussion](https://github.com/git-lfs/git-lfs/discussions) or [issue](https://github.com/git-lfs/git-lfs/issues) and provide as much detail as possible so we can try to help.  In particular, in the new discussion or issue, include the output of `git lfs env` when run inside your Git repository's working tree.
